src/libraries/System.Console/src/System/ConsolePal.iOS.cs

```diff
@@ -1,22 +1,66 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
 using System.IO;
+using System.Runtime.InteropServices;
 using System.Text;
 
 namespace System
 {
     internal sealed class NSLogStream : ConsoleStream
     {
+        private readonly StringBuilder _buffer = new StringBuilder();
+
         public NSLogStream() : base(FileAccess.Write) {}
 
         public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
 
         public override unsafe void Write(ReadOnlySpan<byte> buffer)
         {
-            fixed (byte* ptr = buffer)
+            ReadOnlySpan<char> charSpan = MemoryMarshal.Cast<byte, char>(buffer);
```

Comment:
This doesn't look valid.  Why is it guaranteed that the input bytes are actually two-byte chars?

Reply:
if I understand things correctly this is because we set `OutputEncoding => Encoding.Unicode;` on the `ConsolePal`.

Reply:
Then what happens when someone does:
```C#
Stream s = Console.OpenOutputStream();
byte[] bytes = Encoding.Unicode.GetBytes("s");
s.WriteByte(byte[0]);
s.Writebyte(byte[1]);
```
?

Reply:
Yeah that wouldn't work...
I guess we could override `WriteByte()` and throw an exception?

Reply:
It's not just WriteByte. It's `Write(...)` that doesn't write out on the exact alignment expected.  Seems very fragile.

Reply:
Right, do you think adding a check for `buffer.Length % 2 == 0` would be enough?

Reply:
And then do what if it's not?  Are you suggesting it throw even though it's technically valid use?

Reply:
Yes, I'd throw an exception, not sure what else we could do?

Reply:
(perhaps not optimal for correctness, but) instead of byte->string->byte, can we keep it as byte and use `MemoryStream` for accumulation and then `printer(stream.ToArray(), index etc. ...)`?

Reply:
> not sure what else we could do?

Buffer the remaining bytes for use on the next call?  Presumably by having a Decoder as a field in the stream and running all bytes through that.
```C#
using System.Text;

var decoder = Encoding.Unicode.GetDecoder();
char[] chars = new char[100];
Span<byte> bytes = stackalloc byte[1];
foreach (byte b in Encoding.Unicode.GetBytes("This is a test. This is only a test."))
{
    bytes[0] = b;
    int i = decoder.GetChars(bytes, chars, flush: false);
    if (i > 0)
    {
        Console.WriteLine($"Got {i}: {new string(chars, 0, i)}");
    }
}
```

Reply:
`NSLogStream` is used to create `StreamWriter` in `Console.CreateOutputWriter`:

https://github.com/dotnet/runtime/blob/f1bf5eee390fa118441cae14a157773f1e2d751d/src/libraries/System.Console/src/System/Console.cs#L240-L250

`StreamWriter` just appends all characters to a private buffer and encodes them to bytes on every `Flush` and in case of `Console` it's for every write (autoflush==true):

https://github.com/dotnet/runtime/blob/f1bf5eee390fa118441cae14a157773f1e2d751d/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L485-L487

perhaps instead of modyfing `NSLogStream` we could create a custom type that derives from `StreamWriter` and change its flushing logic (by overriding `Flush`) to do similar thing that @MaximLipnin wrote here:
* take the private buffer
* find next line
* encode it to bytes
* call Write(bytes)
* repeat until the buffer is empty